Replace backslash in INCLUDE paths

atgeirr opened the pull request:
The following addresses #1033 by converting backslashes in include path strings to slashes. This is done after all other processing such as expanding mapped paths (of the `$PATH_KEY` kind), just before creating the boost::filesystem::path object.

I believe it is a better solution than using a conversion script since:
 - requiring such a script is likely to reduce usage by casual users (it doubles the number of commands to remember...)
 - writing such a script in such a way that it only does what it is supposed to do might be hard, the current solution can only affect INCLUDE paths
 - we can (and do) give a warning about this, so that users can choose to make the paths universal (using slashes) but do not require it

@hnil can you verify that this PR solves the observed problem?

andlaus commented:
on a slightly related topic: are there serious -- preferably technical -- objections to getting rid of Boost::filesystem? If no, I offer to have a look what it takes.

joakim-hove commented:
Thank you for adding test. 'll hold my nose and merge this but first:

1. I guess the `foo_target.inc` file has been forgotten from the PR?

2. Please add some keywords in the `foo_target.inc` file and check that they are present in the resulting `Deck` object.

joakim-hove commented:
I am at the top of my head not aware of serious reasons to hold on to Boost FS.

andlaus commented:
Ok, then I'll have a look soon.

atgeirr commented:
The file is already there: `testdata/parser/include/foopath/foo_target.inc` (it's already used by a different test that I copied, replacing slashes with backslashes in other files).  I have added a check that the one keyword in it is actually found (also added this to the existing INCLUDE test).

joakim-hove commented:
> Ok, then I'll have a look soon.

Mind you - my answer was: "I am at the top of my head not aware of serious reasons to hold on to Boost FS." - i.e. there might still be good reasons to keep Boost FS around, which I had forgotten.

joakim-hove commented on opm/parser/eclipse/Parser/tests/ParserTests.cpp:
```diff
@@ -298,6 +299,7 @@ BOOST_AUTO_TEST_CASE( PATHS_with_backslashes ) {
 
     parseContext.update( ParseContext::PARSE_MISSING_INCLUDE , Opm::InputError::THROW_EXCEPTION);
     parser.parseFile( "testdata/parser/PATHSWithBackslashes.data", parseContext );
+    BOOST_CHECK(parser.hasKeyword("OIL"));
```
Ehhh - this just tests that the parser instance recognizes the `OIL` keyword, not that it has actually loaded that keyword into a `Deck` object. The test you need is:

```C++
const auto deck = parser.parseFile( ... );
BOOST_CHECK( deck.hasKeyword("OIL"));
```

atgeirr replied:
Thanks, I did not realize this distinction.

joakim-hove commented:
Instead of special casing the `INCLUDE` handling I would prefer if you added a
```C++
ParserState::ParserState( const ParseContext& context, const std::string& filename) { ... }
```
overload [here](https://github.com/OPM/opm-parser/blob/master/opm/parser/eclipse/Parser/Parser.cpp#L279) then we could (module shell escaping?) actually run `flow` as:

```txt
bash% flow \path\to\deck/with\slash/confusion/ECL.DATA
```
ugly - but to some extent consistent.

atgeirr commented:
> Instead of special casing the INCLUDE handling I would prefer if you added a
> 
> ParserState::ParserState( const ParseContext& context, const std::string& filename) { ... }
> overload here then we could (modulo shell escaping?) actually run flow as:
> 
> bash% flow \path\to\deck/with\slash/confusion/ECL.DATA
> ugly - but to some extent consistent.

I assume then the backslash detection/replacement should happen in that constructor. For the warning, I assume using the ParseState's deck member's message container would be correct?

joakim-hove commented:
> I assume then the backslash detection/replacement should happen in that constructor. 

Yes.

> For the warning, I assume using the ParseState's deck member's message container would be correct?

Yes.

To go completely overboard we could make a `ParseContext` behavior out of this - so that the behavior could be regulated - although this will be somewhat oddball there.

atgeirr commented:
I have tried and failed to make a ParseState::ParseState() constructor overload work for this. The reason is that in the end it is the ParseState::loadFile() method that is used to actually open an INCLUDEd file, not the constructor.

Adding the logic to loadFile() instead, and changing the constructor(s) to call loadFile() would be possible, but there is a subtlety there: loadFile takes a boost::filesystem::path and not a string, so to do the required manipulation there we'd need to extract the string, replace backslashes, then make it a path again.

While doable, it seems to me that the original proposal is far simpler, and less brittle. It does not give the ability to use backslashes in arbitrary paths (only in INCLUDE), but I don't believe that is a big drawback.

I have pushed an update to the test issue as requested. For the other issue discussed above, please advise.

joakim-hove commented:
> [...] The reason is that in the end it is the ParseState::loadFile() method that is used to actually open an INCLUDEd file, not the constructor.

Well - yes; so the constructor would not the intercept in the `std::string -> boost::filesystem::path` transition, and then continue processing with a valid `boost::filesystem::path` instance. Let it be if it is to much hassle.

atgeirr commented:
> Let it be if it is to much hassle.

Appreciated. In that case I think there are no more changes from me here unless you have further requests.

andlaus commented:
> Mind you - my answer was: "I am at the top of my head not aware of serious reasons to hold on to Boost FS." - i.e. there might still be good reasons to keep Boost FS around, which I had forgotten.

Ok, I had a look: there seems to be a reason to keep it, mainly having a way iterate over directories without resorting to operating system specific low-level functions. That said it looks like about 90% of the users of boost::filesystem in opm-parsers do not really seem to need...

andlaus commented:
note that boost::filesystem recently transformed into std::filesystem. unfortunately its only available from C++-2017 on...

atgeirr commented:
> std::filesystem

Hooray! What really matters most for us is not necessarily the C++ version, but which gcc version that includes it (and when that version is the default or available on our platforms). Do you know if gcc 5 has this?

andlaus commented:
I googled a bit for it and it seems like GCC 5 does not include it at all while GCC 6 has `std::experimental::filesystem`. looks like we have to wait a few years ;)
